### PR TITLE
RYA-53 add Visibilty Support to RYA

### PR DESCRIPTION
**Pull request by isper3at**
Added Visibility support for PCJs and fluo with integration tests.
Added a BindingSet decorator.

[Jira](https://issues.apache.org/jira/browse/RYA-53)

### CODE REVIEW

File: extras/indexing/src/main/java/mvm/rya/indexing/external/tupleSet/BindingSetConverter.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Licensed to the Apache Software Foundation (ASF) under one
+* Licensed to the Apache Software Foundation (ASF) under one
```
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/indexing/external/tupleSet/BindingSetConverter.java]: remove this

File: extras/indexing/src/main/java/mvm/rya/indexing/external/tupleSet/FullBindingSetStringConverter.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import mvm.rya.api.domain.RyaType;
+import mvm.rya.api.resolver.RdfToRyaConversions;
+import mvm.rya.indexing.external.tupleSet.PcjTables.VariableOrder;
+
+/**
+ * Converts {@link BindingSet}s to Strings and back again. The Strings do not
+ * include the binding names and are ordered with a {@link VariableOrder}.
+ */
+@ParametersAreNonnullByDefault
+public class FullBindingSetStringConverter extends BindingSetStringConverter {
```
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/indexing/external/tupleSet/FullBindingSetStringConverter.java]: Rename?

File: extras/indexing/src/main/java/mvm/rya/indexing/accumulo/VisibilityBindingSet.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import org.openrdf.query.BindingSet;
+
+import mvm.rya.indexing.external.BindingSetDecorator;
+
+/**
+ * Decorates a {@link BindingSet} with a collection of visibilities.
+ */
```
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/indexing/accumulo/VisibilityBindingSet.java]: add parametersAreNonNullableByDefault

File: extras/indexing/src/test/java/mvm/rya/indexing/external/tupleSet/PcjTablesIntegrationTests.java [hunk 4]
```diff
@@ -151,35 +151,39 @@ public void addResults() throws PcjException, TableNotFoundException, BindingSet
...
         bob.addBinding("age", new NumericLiteralImpl(16, XMLSchema.INTEGER));
 
-        MapBindingSet charlie = new MapBindingSet();
+        final MapBindingSet charlie = new MapBindingSet();
         charlie.addBinding("name", new URIImpl("http://Charlie"));
         charlie.addBinding("age", new NumericLiteralImpl(12, XMLSchema.INTEGER));
 
-        Set<BindingSet> results = Sets.<BindingSet>newHashSet(alice, bob, charlie);
+        final Set<VisibilityBindingSet> results = Sets.<VisibilityBindingSet>newHashSet(
+                new VisibilityBindingSet(alice),
```
Comment (isper3at) [inline extras/indexing/src/test/java/mvm/rya/indexing/external/tupleSet/PcjTablesIntegrationTests.java]: whitespace

File: extras/rya.pcj.fluo/pcj.fluo.api/src/main/java/org/apache/rya/indexing/pcj/fluo/api/InsertTriples.java [hunk 5]
```diff
@@ -53,30 +55,35 @@
      */
     private static final WholeRowTripleResolver TRIPLE_RESOLVER = new WholeRowTripleResolver();
 
+    private static final Encoder encoder = new StringEncoder();
```
Comment (isper3at) [inline extras/rya.pcj.fluo/pcj.fluo.api/src/main/java/org/apache/rya/indexing/pcj/fluo/api/InsertTriples.java]: ENCODER

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/FilterResultUpdater.java [hunk 6]
```diff
@@ -62,7 +64,8 @@
 
     private final Encoder encoder = new StringEncoder();
 
-    private final BindingSetStringConverter converter = new BindingSetStringConverter();
+    private final BindingSetStringConverter idConverter = new BindingSetStringConverter();
+    private final VisibilityBindingSetStringConverter valueConverter = new VisibilityBindingSetStringConverter();
```
Comment (isper3at) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/FilterResultUpdater.java]: static, [some]_CONVERTER.  Might as well update ENCODER while here

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/observers/QueryResultObserver.java [hunk 7]
```diff
@@ -48,8 +50,8 @@
     private static final Logger log = Logger.getLogger(QueryResultObserver.class);
 
     private final FluoQueryMetadataDAO queryDao = new FluoQueryMetadataDAO();
-
-    private final BindingSetStringConverter converter = new BindingSetStringConverter();
+    private final Encoder encoder = new StringEncoder();
+    private final VisibilityBindingSetStringConverter converter = new VisibilityBindingSetStringConverter();
```
Comment (isper3at) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/observers/QueryResultObserver.java]: static, CONVERTER, ENCODER, might as well update QUERY_DAO?

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/observers/TripleObserver.java [hunk 8]
```diff
@@ -50,6 +55,8 @@
  */
 public class TripleObserver extends TypedObserver {
 
+    private static final Encoder ENCODER = new StringEncoder();
+
```
Comment (isper3at) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/observers/TripleObserver.java]: Update QUERY_DAO?

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/observers/TripleObserver.java [hunk 9]
```diff
@@ -65,6 +72,12 @@ public ObservedColumn getObservedColumn() {
     public void process(final TypedTransactionBase tx, final Bytes row, final Column column) {
         //get string representation of triple
         final String triple = IncUpdateDAO.getTripleString(row);
+        final Bytes visiBytes = tx.get(row, FluoQueryColumns.TRIPLES);
+        String visibility = "";
+        if(visiBytes != null) {
+             visibility = ENCODER.decodeString(visiBytes);
+        }
+        final VisibilityBindingSetStringConverter converter = new VisibilityBindingSetStringConverter();
```
Comment (isper3at) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/observers/TripleObserver.java]: move out

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/observers/TripleObserver.java [hunk 10]
```diff
@@ -85,9 +97,16 @@ public void process(final TypedTransactionBase tx, final Bytes row, final Column
 
             while(ci.hasNext()) {
                 final String varOrders = ci.next().getValue().toString();
-                final String bindingSet = getBindingSet(triple, pattern, varOrders);
-                if(bindingSet.length() != 0) {
-                    tx.mutate().row(spID + NODEID_BS_DELIM + bindingSet).col(FluoQueryColumns.STATEMENT_PATTERN_BINDING_SET).set(bindingSet);
+                final VariableOrder varOrder = new VariableOrder(varOrders);
+                final String bindingSetString = getBindingSet(triple, pattern, varOrders);
+
```
Comment (isper3at) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/observers/TripleObserver.java]: whitespace

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/observers/TripleObserver.java [hunk 11]
```diff
@@ -96,8 +115,14 @@ public void process(final TypedTransactionBase tx, final Bytes row, final Column
         tx.delete(row, column);
     }
 
-    //determines whether triple matches SPID conditions and generates bindingset
-    //whose order is determined by varOrder
+    /**
+     * Determines whether triple matches SPID conditions and generates bindingset
```
Comment (isper3at) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/observers/TripleObserver.java]: clean up docs

File: extras/rya.pcj.fluo/pcj.fluo.client/src/main/java/org/apache/rya/indexing/pcj/fluo/client/util/FluoLoader.java [hunk 12]
```diff
@@ -68,7 +68,7 @@ public void handleStatement(final Statement st) throws RDFHandlerException {
         // If the buffer is full, flush it to the Fluo table.
         if(buff.size() == FLUSH_SIZE) {
             log.trace("Flushing " + buff.size() + " Statements from the buffer to Fluo.");
-            insertTriples.insert(fluoClient, buff);
+            insertTriples.insert(fluoClient, buff, "");
```
Comment (isper3at) [inline extras/rya.pcj.fluo/pcj.fluo.client/src/main/java/org/apache/rya/indexing/pcj/fluo/client/util/FluoLoader.java]: overload insert so empty quotes not necessary

File: extras/rya.pcj.fluo/pcj.fluo.app/src/test/java/org/apache/rya/indexing/pcj/fluo/app/NaturalJoinTest.java [hunk 13]
```diff
@@ -46,120 +48,132 @@
...
 
-        MapBindingSet nameHeightHair = new MapBindingSet();
+        final MapBindingSet nameHeightHair = new MapBindingSet();
         nameHeightHair.addBinding("name", vf.createLiteral("Bob"));
         nameHeightHair.addBinding("height", vf.createLiteral("5'9\""));
         nameHeightHair.addBinding("hairColor", vf.createLiteral("Brown"));
-        expected.add(nameHeightHair);
+        expected.add(new VisibilityBindingSet(nameHeightHair));
+
+        for(final BindingSet bs : newJoinResults) {
```
Comment (kchilton2) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/test/java/org/apache/rya/indexing/pcj/fluo/app/NaturalJoinTest.java]: This debug code should be removed.

File: extras/indexing/src/main/java/mvm/rya/indexing/accumulo/VisibilityBindingSet.java [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+            return true;
+        } else if(o instanceof VisibilityBindingSet) {
+            final VisibilityBindingSet other = (VisibilityBindingSet) o;
+            return set.equals(other) && visibility.equals(other.getVisibility());
+        }
+        return false;
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/indexing/accumulo/VisibilityBindingSet.java]: Do we invoke hashCode often enough that we want to cache the value?

File: extras/indexing/src/main/java/mvm/rya/indexing/external/BindingSetDecorator.java [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * Abstracts out the decoration of a {@link BindingSet}.
+ */
+public abstract class BindingSetDecorator implements BindingSet {
+    private static final long serialVersionUID = 1L;
+    protected final BindingSet set;
+    private volatile int hashCode;
+
+    public BindingSetDecorator(final BindingSet set) {
```
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/indexing/external/BindingSetDecorator.java]: What is the nullness contract here? Also missing docs.

File: extras/indexing/src/main/java/mvm/rya/indexing/external/BindingSetDecorator.java [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+    @Override
+    public int size() {
+        return set.size();
+    }
+
+    @Override
+    public int hashCode() {
+        int result = hashCode;
+        if(result == 0) {
+            result = 31 * result + size();
```
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/indexing/external/BindingSetDecorator.java]: Why do you include the size of the binding set as part of the hash code?

File: extras/indexing/src/main/java/mvm/rya/indexing/external/tupleSet/FullBindingSetStringConverter.java (hunk 2 above)
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/indexing/external/tupleSet/FullBindingSetStringConverter.java]: Because this implementation totally converts binding sets (including the binding names), you don't need to use variable order objects to keep track of the bindings.

File: extras/indexing/src/test/java/mvm/rya/indexing/external/tupleSet/VisibilityBindingSetStringConverterTest.java [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+        final VariableOrder varOrder = new VariableOrder("y", "z", "x");
+        final VisibilityBindingSetStringConverter converter = new VisibilityBindingSetStringConverter();
+        final String bindingSetString = converter.convert(visiSet, varOrder);
+
+        // Ensure it converted to the expected result.l
+        final String expected =
+                "http://b<<~>>http://www.w3.org/2001/XMLSchema#anyURI:::" +
+                "http://c<<~>>http://www.w3.org/2001/XMLSchema#anyURI:::" +
+                "http://a<<~>>http://www.w3.org/2001/XMLSchema#anyURI" +
+                (char)1 + "A&B&C";
```
Comment (kchilton2) [inline extras/indexing/src/test/java/mvm/rya/indexing/external/tupleSet/VisibilityBindingSetStringConverterTest.java]: Might want to store the (char)1 delimiter as a constant somewhere and reference that here.

File: extras/rya.pcj.fluo/pcj.fluo.api/src/main/java/org/apache/rya/indexing/pcj/fluo/api/CreatePcj.java [hunk 18]
```diff
@@ -139,6 +141,7 @@ public CreatePcj(final int spInsertBatchSize) {
      * @throws QueryEvaluationException Historic results could not be added to the initialized PCJ because of
      *   a problem with the Rya connection.
      */
+    @VisibleForTesting
```
Comment (kchilton2) [inline extras/rya.pcj.fluo/pcj.fluo.api/src/main/java/org/apache/rya/indexing/pcj/fluo/api/CreatePcj.java]: Why was this added?

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/FilterResultUpdater.java [hunk 19]
```diff
@@ -126,11 +129,19 @@ public void updateFilterResults(
                     filterBindingSet.addBinding(binding);
                 }
             }
-            final String filterBindingSetString = converter.convert(filterBindingSet, filterVarOrder);
 
-            final Bytes row = encoder.encode( filterMetadata.getNodeId() + NODEID_BS_DELIM + filterBindingSetString );
+            final String filterBindingSetIdString = ID_CONVERTER.convert(filterBindingSet, filterVarOrder);
+            String filterBindingSetValueString = "";
+            if(childBindingSet instanceof VisibilityBindingSet) {
```
Comment (kchilton2) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/FilterResultUpdater.java]: We shouldn't do the type checking here. Assume it's always a VisibilityBindingSet once it's within the Fluo app. Lets update the BindingSet object that is passed into the updaters to require a VisibilityBindingSet.

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/JoinResultUpdater.java [hunk 20]
```diff
@@ -407,15 +411,35 @@ public boolean hasNext() {
...
 
-            for(Binding binding : joinedResults.next()) {
+            final BindingSet joinResult = joinedResults.next();
+            for(final Binding binding : joinResult) {
                 bs.addBinding(binding);
             }
 
-            return bs;
+            String visibility = "";
+            if(newResult instanceof VisibilityBindingSet && joinResult instanceof VisibilityBindingSet) {
```
Comment (kchilton2) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/JoinResultUpdater.java]: This is another area where we should be able to assume we're using VisibilityBindingSets.

File: extras/rya.pcj.fluo/pcj.fluo.app/src/test/java/org/apache/rya/indexing/pcj/fluo/app/NaturalJoinTest.java (hunk 13 above)
Comment (isper3at) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/test/java/org/apache/rya/indexing/pcj/fluo/app/NaturalJoinTest.java]: hah woops.  done

File: extras/indexing/src/main/java/mvm/rya/indexing/accumulo/VisibilityBindingSet.java (hunk 14 above)
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/indexing/accumulo/VisibilityBindingSet.java]: probably not a bad call to cache it anyway

File: extras/indexing/src/main/java/mvm/rya/indexing/external/BindingSetDecorator.java (hunk 15 above)
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/indexing/external/BindingSetDecorator.java]: done

File: extras/indexing/src/main/java/mvm/rya/indexing/external/BindingSetDecorator.java (hunk 16 above)
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/indexing/external/BindingSetDecorator.java]: I was initially going through the fields, it didn't occur to me at the time that it was a changing value

File: extras/indexing/src/main/java/mvm/rya/indexing/external/tupleSet/FullBindingSetStringConverter.java (hunk 2 above)
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/indexing/external/tupleSet/FullBindingSetStringConverter.java]: huh....I was using it...I deleted it and everything still works sooooo....huh...

File: extras/indexing/src/test/java/mvm/rya/indexing/external/tupleSet/VisibilityBindingSetStringConverterTest.java (hunk 17 above)
Comment (isper3at) [inline extras/indexing/src/test/java/mvm/rya/indexing/external/tupleSet/VisibilityBindingSetStringConverterTest.java]: done

File: extras/rya.pcj.fluo/pcj.fluo.api/src/main/java/org/apache/rya/indexing/pcj/fluo/api/CreatePcj.java (hunk 18 above)
Comment (isper3at) [inline extras/rya.pcj.fluo/pcj.fluo.api/src/main/java/org/apache/rya/indexing/pcj/fluo/api/CreatePcj.java]: artifact left over from when I added the withRyaIntegration() that returned the generated table name

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/FilterResultUpdater.java (hunk 19 above)
Comment (isper3at) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/FilterResultUpdater.java]: I'll go through and update the interfaces?

File: extras/indexing/src/main/java/mvm/rya/indexing/external/BindingSetDecorator.java [hunk 21]
```diff
@@ -71,7 +78,6 @@ public int size() {
     public int hashCode() {
         int result = hashCode;
```
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/indexing/external/BindingSetDecorator.java]: Why not just return the set's hash code?

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/JoinResultUpdater.java [hunk 22]
```diff
@@ -408,38 +408,28 @@ public boolean hasNext() {
...
-                visibility = ((VisibilityBindingSet) joinResult).getVisibility();
-                return new VisibilityBindingSet(bs, visibility);
+            final Joiner join = Joiner.on(")&(");
+            final String leftVisi = newResult.getVisibility();
+            final String rightVisi = joinResult.getVisibility();
+            if(leftVisi.isEmpty() || rightVisi.isEmpty()) {
+                visibility = (leftVisi + rightVisi).trim();
             } else {
-                return new VisibilityBindingSet(bs);
+                visibility = "(" + join.join(leftVisi, rightVisi) + ")";
```
Comment (kchilton2) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/JoinResultUpdater.java]: Using a joiner here is kind of overkill since you're only ever joining two items.

File: extras/indexing/src/main/java/mvm/rya/indexing/external/tupleSet/FullBindingSetStringConverter.java (hunk 2 above)
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/indexing/external/tupleSet/FullBindingSetStringConverter.java]: The original idea was to store the binding set + visibility within the value section so that you would no longer need to unmarshal the row id using the variable order, but I don't think you're using that approach.

File: extras/indexing/src/main/java/mvm/rya/indexing/external/BindingSetDecorator.java [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+    @Override
+    public Value getValue(final String bindingName) {
+        return set.getValue(bindingName);
+    }
+
+    @Override
+    public int size() {
+        return set.size();
+    }
+
```
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/indexing/external/BindingSetDecorator.java]: You should implement an equals method since you're overriding hashcode.

File: extras/indexing/src/main/java/mvm/rya/indexing/external/tupleSet/BindingSetStringConverter.java [hunk 24]
```diff
@@ -57,48 +57,48 @@
...
-     * 
+     *
      * @param bindingSet - The binding set whose Bindings will be inspected. (not null)
      * @param varOrder - The names of the bindings that may appear in the BindingSet. (not null)
      * @throws IllegalArgumentException Indicates the names of the bindings are
-     *   not a subset of the variable order. 
+     *   not a subset of the variable order.
      */
-    private static void checkBindingsSubsetOfVarOrder(BindingSet bindingSet, VariableOrder varOrder) throws IllegalArgumentException {
+    protected static void checkBindingsSubsetOfVarOrder(final BindingSet bindingSet, final VariableOrder varOrder) throws IllegalArgumentException {
```
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/indexing/external/tupleSet/BindingSetStringConverter.java]: You made this projected, but never call it outside of this class. It should remain private.

File: extras/indexing/src/main/java/mvm/rya/indexing/external/tupleSet/VisibilityBindingSetStringConverter.java [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+ * include the binding names and are ordered with a {@link VariableOrder}.
+ */
+@ParametersAreNonnullByDefault
+public class VisibilityBindingSetStringConverter extends BindingSetStringConverter {
+    public static final char VISIBILITY_DELIM = 1;
+
+    @Override
+    public String convert(final BindingSet bindingSet, final VariableOrder varOrder) {
+        String visibility = "";
+        if(bindingSet instanceof VisibilityBindingSet) {
```
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/indexing/external/tupleSet/VisibilityBindingSetStringConverter.java]: Using instanceof here is kind of an abstraction smell, but I'm not sure of how to fix this other than maybe having each converter take a generic type that extends BindingSet.

Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/indexing/external/tupleSet/VisibilityBindingSetStringConverter.java]: each converter already takes a generic that is what the binding set is being converted to/from

**Comment by pujav65** [comment]
no comments on this -- i will merge it once i verify tests pass.

**Comment by pujav65** [comment]
merged.  can you please close the pull request?
